zkevm-circuits/src/evm_circuit/op_execution/swap.rs

```diff
@@ -0,0 +1,482 @@
+            ),
+        }
+    }
+
+    fn constraints(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>> {
+        let OpExecutionState { opcode, .. } = &state_curr;
+
+        let common_lookups = vec![Lookup::FixedLookup(
+            FixedLookup::Range16,
+            [opcode.expr() - OpcodeId::SWAP1.expr(), 0.expr(), 0.expr()],
+        )];
+
+        let num_swaped = opcode.expr() - OpcodeId::SWAP1.expr() + 1.expr();
+        let success = {
+            let SwapSuccessAllocation {
+                case_selector,
+                words,
+            } = &self.success;
+
+            let state_transition_constraints = vec![
+                state_next.global_counter.expr()
+                    - (state_curr.global_counter.expr() + 4.expr()),
+                state_next.program_counter.expr()
+                    - (state_curr.program_counter.expr() + 1.expr()),
+                state_next.gas_counter.expr()
+                    - (state_curr.gas_counter.expr() + GasCost::FASTEST.expr()),
+            ];
+
+            /*
+                bus_mapping_lookup(gc, Stack, key+X, $val1, Read)
+                bus_mapping_lookup(gc+1, Stack, key, $val2, Read)
+                bus_mapping_lookup(gc+2, Stack, key+X, $val2, Write)
+                bus_mapping_lookup(gc+3, Stack, key, $val1, Write)
+            */
+
+            let bus_mapping_lookups = vec![
+                // constrain (num_swaped)th value of stack
+                Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: num_swaped.clone(),
+                    value: words[0].expr(),
+                    is_write: false,
+                }),
+                // constrain top value of stack
+                Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: 0.expr(),
+                    value: words[1].expr(),
+                    is_write: false,
+                }),
+                // constrains when swap done
+                Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: num_swaped.clone(),
+                    value: words[1].expr(),
+                    is_write: true,
+                }),
+                Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: 0.expr(),
+                    value: words[0].expr(),
+                    is_write: true,
+                }),
+            ];
+
+            Constraint {
+                name: "SwapGadget success",
+                selector: case_selector.expr(),
+                polys: state_transition_constraints,
+                lookups: bus_mapping_lookups,
+            }
+        };
+
+        // stack_underflow condition: 0 <= num_swaped +  stack_pointer - 1025 < 16
+        let stack_pointer = state_curr.stack_pointer.expr();
+        let diff = num_swaped + stack_pointer - 1025.expr();
```

Comment:
The stack underflow case for `SWAP*` is slightly different than `DUP*`, we need `stack.len() >= num_swaped + 1`, for example the `SWAP1` swaps `stack[-1]` and `stack[-2]`.

So the range lookup should be `0 <= num_swaped +  stack_pointer - 1024 < 17`, notice that the max value is `16` dut to `SWAP16`. So we might probably need a `Range17` tag for this.